File: src/main/java/sh/ball/audio/effect/BulgeEffect.java
package sh.ball.audio.effect;

import sh.ball.shapes.Vector2;

public class BulgeEffect implements SettableEffect {

  private double bulge = 0.1;

  @Override
  public Vector2 apply(int count, Vector2 vector) {
    double translatedBulge = -bulge + 1;

    double r = vector.magnitude();
    double theta = Math.atan2(vector.y, vector.x);
    double rn = Math.pow(r, translatedBulge);

    return new Vector2(rn * Math.cos(theta), rn * Math.sin(theta));
  }

  @Override
  public void setValue(double value) {
    this.bulge = value;
  }
}

File: Source/audio/BulgeEffect.cpp
#include "BulgeEffect.h"

BulgeEffect::BulgeEffect() {}

BulgeEffect::~BulgeEffect() {}

OsciPoint BulgeEffect::apply(int index, OsciPoint input, const std::vector<std::atomic<double>>& values, double sampleRate) {
    double value = values[0];
    double translatedBulge = -value + 1;

    double r = sqrt(pow(input.x, 2) + pow(input.y, 2));
    double theta = atan2(input.y, input.x);
    double rn = pow(r, translatedBulge);

    return OsciPoint(rn * cos(theta), rn * sin(theta), input.z);
}
